src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncSpillTests.cs

```diff
@@ -3888,5 +3887,2228 @@ class AltBoolean
+    IL_006b:  stfld      ""int Program.<Assign>d__0.<>1__state""
+    IL_0070:  ldloca.s   V_2
+    IL_0072:  call       ""int System.Runtime.CompilerServices.TaskAwaiter<int>.GetResult()""
+    IL_0077:  stloc.1
+    IL_0078:  ldarg.0
+    IL_0079:  ldfld      ""B Program.<Assign>d__0.<>7__wrap1""
+    IL_007e:  ldloc.1
+    IL_007f:  stfld      ""int B.x""
+    IL_0084:  ldarg.0
+    IL_0085:  ldnull
+    IL_0086:  stfld      ""B Program.<Assign>d__0.<>7__wrap1""
+    IL_008b:  leave.s    IL_00a4
+  }
+  catch System.Exception
+  {
+    IL_008d:  stloc.3
+    IL_008e:  ldarg.0
+    IL_008f:  ldc.i4.s   -2
+    IL_0091:  stfld      ""int Program.<Assign>d__0.<>1__state""
+    IL_0096:  ldarg.0
+    IL_0097:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder Program.<Assign>d__0.<>t__builder""
+    IL_009c:  ldloc.3
+    IL_009d:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetException(System.Exception)""
+    IL_00a2:  leave.s    IL_00b7
+  }
+  IL_00a4:  ldarg.0
+  IL_00a5:  ldc.i4.s   -2
+  IL_00a7:  stfld      ""int Program.<Assign>d__0.<>1__state""
+  IL_00ac:  ldarg.0
+  IL_00ad:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder Program.<Assign>d__0.<>t__builder""
+  IL_00b2:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()""
+  IL_00b7:  ret
+}");
+        }
+
+        [Fact]
+        [WorkItem(42755, "https://github.com/dotnet/roslyn/issues/42755")]
+        public void KeepLtrSemantics_ClassFieldAccessOnArray()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Program
+{
+    static async Task Assign(A[] arr)
+    {
+        arr[0].x = await Write(""RHS"");
+    }
+
+    static async Task Main(string[] args)
+    {
+        await TestIndexerThrows();
+        await TestAssignmentThrows();
+        await TestIndexerSucceeds();
+        await TestReassignsArrayAndIndexerDuringAwait();
+        await TestReassignsTargetDuringAwait();
+    }
+
+    static async Task TestIndexerThrows()
+    {
+        Console.WriteLine(nameof(TestIndexerThrows));
+        
+        var arr = new A[0];
+        Console.WriteLine(""Before Assignment"");
+        try
+        {
+            await Assign(arr);
+        }
+        catch (IndexOutOfRangeException)
+        {
+            Console.WriteLine(""Caught IndexOutOfRangeException"");
+        }
+    }
+
+    static async Task TestAssignmentThrows()
+    {
+        Console.WriteLine(nameof(TestAssignmentThrows));
+        
+        var arr = new A[1];
+        Console.WriteLine(""Before Assignment"");
+        try
+        {
+            await Assign(arr);
+        }
+        catch (NullReferenceException)
+        {
+            Console.WriteLine(""Caught NullReferenceException"");
+        }
+    }
+
+    static async Task TestIndexerSucceeds()
+    {
+        Console.WriteLine(nameof(TestIndexerSucceeds));
+
+        var arr = new A[1]{ new A() };
```

Comment:
>new A() [](start = 28, length = 7)

Is it interesting to test scenario when this is null?

Reply:
if you mean the element, then that is the test at https://github.com/dotnet/roslyn/pull/42807/files/850e04de53a20e21f648b7cd2cb7f87ad1b16371#diff-e671cb4c1486699ffe0e0bd95d1a4ff5R4110
If you mean the array, I could add that test, although it is pretty much the same as the test starting at https://github.com/dotnet/roslyn/pull/42807/files/850e04de53a20e21f648b7cd2cb7f87ad1b16371#diff-e671cb4c1486699ffe0e0bd95d1a4ff5R4094